toolchain/check/impl.cpp

```diff
@@ -107,33 +107,52 @@ auto ImplWitnessStartDefinition(Context& context, SemIR::Impl& impl) -> void {
   if (impl.witness_id == SemIR::ErrorInst::InstId) {
     return;
   }
+
+  if (!RequireCompleteType(
+          context, context.types().GetTypeIdForTypeInstId(impl.constraint_id),
+          SemIR::LocId(impl.constraint_id), [&] {
+            CARBON_DIAGNOSTIC(ImplAsIncompleteFacetTypeDefinition, Error,
+                              "definition of impl as incomplete facet type {0}",
+                              InstIdAsType);
+            return context.emitter().Build(SemIR::LocId(impl.latest_decl_id()),
+                                           ImplAsIncompleteFacetTypeDefinition,
+                                           impl.constraint_id);
+          })) {
+    FillImplWitnessWithErrors(context, impl);
+    return;
+  }
+
+  const auto& interface = context.interfaces().Get(impl.interface.interface_id);
+
+  auto assoc_entities =
+      context.inst_blocks().Get(interface.associated_entities_id);
+  for (auto decl_id : assoc_entities) {
+    LoadImportRef(context, decl_id);
+  }
+
   auto witness = context.insts().GetAs<SemIR::ImplWitness>(impl.witness_id);
   auto witness_table =
       context.insts().GetAs<SemIR::ImplWitnessTable>(witness.witness_table_id);
   auto witness_block =
       context.inst_blocks().GetMutable(witness_table.elements_id);
-  // `witness_table.elements_id` will be `SemIR::InstBlockId::Empty` when the
-  // definition is the first declaration and the interface has no members. The
-  // other case where `witness_block` will be empty is when we are using a
-  // placeholder witness. This happens when there is a forward declaration of
-  // the impl and the facet type has no rewrite constraints and so it wasn't
-  // required to be complete.
-  if (witness_table.elements_id != SemIR::InstBlockId::Empty &&
-      witness_block.empty()) {
-    if (!RequireCompleteFacetTypeForImplDefinition(
-            context, SemIR::LocId(impl.latest_decl_id()), impl.constraint_id)) {
-      FillImplWitnessWithErrors(context, impl);
-      return;
-    }
 
-    AllocateFacetTypeImplWitness(context, impl.interface.interface_id,
-                                 witness_table.elements_id);
+  // We can detect that the witness table is a placeholder table if it's not the
+  // `Empty` id, but it is empty still. This happens when a forward declaration
+  // makes the table before the interface is known to be complete (so the
+  // table's required size was unknown). Then we replace the placeholder table
+  // here in the definition, as the interface must be complete.
+  bool witness_table_is_placeholder =
+      witness_table.elements_id != SemIR::InstBlockId::Empty &&
+      witness_block.empty();
+  if (witness_table_is_placeholder) {
+    // TODO: We could skip an allocation here if we could `ReplacePlaceholder`
+    // with a size and value.
```

Comment:
Could you clarify this TODO? I think you mean add an overload to `ReplacePlaceholder` that takes a size and value instead of a vector of `InstId`s. Or are you saying all uses of `ReplacePlaceholder` would use that API?